File: backend/src/main/java/unpsjb/labprog/backend/business/service/LicenciaService.java
package unpsjb.labprog.backend.business.service;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import jakarta.transaction.Transactional;
import unpsjb.labprog.backend.business.repository.LicenciaRepository;
import unpsjb.labprog.backend.business.validator.licencia.LicenciaValidator;
import unpsjb.labprog.backend.exception.BusinessLogicException;
import unpsjb.labprog.backend.exception.NotModifiableException;
import unpsjb.labprog.backend.model.Designacion;
import unpsjb.labprog.backend.model.Licencia;
import unpsjb.labprog.backend.model.Log;
import unpsjb.labprog.backend.model.Persona;
import unpsjb.labprog.backend.model.enums.Estado;

/**
 * Servicio para la gestión de licencias de personal docente. Proporciona
 * operaciones para crear, validar, consultar y administrar licencias,
 * incluyendo la validación de reglas de negocio y gestión de estados.
 *
 * @author Mariano Bonansea
 * @version 1.0
 */
@Service
public class LicenciaService {

    @Autowired
    private LicenciaRepository repository;

    @Autowired
    private LicenciaValidator validator;

    @Autowired
    private DesignacionService designacionService;

    /**
     * Obtiene todas las licencias registradas en el sistema.
     *
     * @return lista de todas las licencias
     */
    public List<Licencia> findAll() {
        return repository.findAll();
    }

    /**
     * Busca una licencia por su identificador único.
     *
     * @param id identificador de la licencia
     * @return licencia encontrada o null si no existe
     */
    public Licencia findById(int id) {
        return repository.findById(id).orElse(null);
    }

    /**
     * Guarda una licencia aplicando validaciones de reglas de negocio. El
     * estado de la licencia se establece automáticamente según el resultado de
     * la validación. También se vinculan las designaciones afectadas por el
     * período de la licencia.
     *
     * @param licencia licencia a guardar
     * @return licencia guardada con estado actualizado
     * @throws NotModifiableException si se intenta modificar una licencia ya
     * validada
     */
    public Licencia save(Licencia licencia) throws NotModifiableException {
        if (licencia.getEstado() != null && licencia.getEstado() == Estado.VALIDO) {
            throw new NotModifiableException("No se puede modificar una licencia ya validada");
        }

        try {
            validator.validar(licencia);

            licencia.setEstado(Estado.VALIDO);

            List<Designacion> designacionesAfectadas = designacionService.findDesignacionesActivasPorPersonaYPeriodo(
                    licencia.getPersona().getDni(),
                    licencia.getPedidoDesde(),
                    licencia.getPedidoHasta());

            Log logExito = new Log();
            logExito.setFechaHora(LocalDateTime.now());
            logExito.setDescripcion("Licencia validada correctamente");
            licencia.getLogs().add(logExito);

            licencia.setDesignaciones(designacionesAfectadas);
        } catch (BusinessLogicException e) {
            licencia.setEstado(Estado.INVALIDO);

            Log logError = new Log();
            logError.setFechaHora(LocalDateTime.now());
            logError.setDescripcion(e.getMessage());
            licencia.getLogs().add(logError);
        }

        return repository.save(licencia);
    }

    /**
     * Elimina una licencia del sistema.
     *
     * @param id identificador de la licencia a eliminar
     */
    @Transactional
    public void delete(int id) {
        repository.deleteById(id);
    }

    /**
     * Obtiene una página de licencias con paginación y ordenamiento
     * personalizado. Valida que el campo de ordenamiento sea permitido por
     * seguridad.
     *
     * @param page índice de página (basado en cero)
     * @param size tamaño de la página
     * @param sortField campo por el cual ordenar
     * @param sortDirection dirección del ordenamiento (asc o desc)
     * @return página de licencias con el ordenamiento especificado
     */
    public Page<Licencia> findByPage(int page, int size, String sortField, String sortDirection) {
        String[] allowedFields = {"id", "persona.dni", "pedidoDesde", "pedidoHasta", "certificadoMedico",
            "articuloLicencia.articulo", "estado"};
        boolean isValidField = false;
        for (String field : allowedFields) {
            if (field.equals(sortField)) {
                isValidField = true;
                break;
            }
        }

        if (!isValidField) {
            sortField = "id";
        }

        Sort.Direction direction;
        if ("asc".equalsIgnoreCase(sortDirection)) {
            direction = Sort.Direction.ASC;
        } else {
            direction = Sort.Direction.DESC;
        }

        return repository.findAll(PageRequest.of(page, size, Sort.by(direction, sortField)));
    }

    /**
     * Busca licencias válidas de una persona en un año específico. Utilizado
     * principalmente para generar reportes.
     *
     * @param persona persona asociada a las licencias
     * @param anio año a consultar
     * @return lista de licencias válidas de la persona en el año especificado
     */
    public List<Licencia> findLicenciasPorPersonaYAño(Persona persona, Integer anio) {
        return repository.findLicenciasPorPersonaYAño(persona, anio);
    }
}

File: frontend/cli/src/app/licencia/service/licencia.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { DataPackage } from '../../models/data-package';
import { Licencia } from '../../models/licencia';

/**
 * Servicio para gestión de licencias y partes diarios del sistema.
 * 
 * Proporciona operaciones CRUD completas para licencias docentes,
 * además de funcionalidades especializadas para generación de partes
 * diarios de licencias por fecha específica.
 * 
 * @author Mariano Bonansea
 * @version 1.0
 */
@Injectable({
    providedIn: 'root'
})
export class LicenciaService {
    private licenciasUrl = "/rest/licencias";

    /**
     * Constructor del servicio.
     * 
     * @param http Cliente HTTP de Angular para realizar peticiones al backend
     */
    constructor(private http: HttpClient) { }

    /**
     * Obtiene todas las licencias del sistema.
     * 
     * @returns Observable con el paquete de datos conteniendo la lista completa de licencias
     */
    all(): Observable<DataPackage> {
        return this.http.get<DataPackage>(encodeURI(this.licenciasUrl));
    }

    /**
     * Obtiene una licencia específica por su ID.
     * 
     * @param id Identificador único de la licencia
     * @returns Observable con el paquete de datos conteniendo la licencia solicitada
     */
    get(id: number): Observable<DataPackage> {
        return this.http.get<DataPackage>(encodeURI(`${this.licenciasUrl}/${id}`));
    }

    /**
     * Guarda una licencia en el sistema (creación o actualización).
     * 
     * @param licencia Objeto licencia a guardar
     * @param isNew Indica si es una creación (true) o actualización (false)
     * @returns Observable con el resultado de la operación
     */
    save(licencia: Licencia, isNew: boolean = false): Observable<DataPackage> {
        return isNew
            ? this.http.post<DataPackage>(encodeURI(this.licenciasUrl), licencia)
            : this.http.put<DataPackage>(encodeURI(this.licenciasUrl), licencia);
    }

    /**
     * Elimina una licencia del sistema.
     * 
     * @param id Identificador único de la licencia a eliminar
     * @returns Observable con el resultado de la operación de eliminación
     */
    remove(id: number): Observable<DataPackage> {
        return this.http.delete<DataPackage>(encodeURI(`${this.licenciasUrl}/${id}`));
    }

    /**
     * Obtiene licencias paginadas con ordenamiento personalizable.
     * 
     * @param page Número de página (base 1)
     * @param size Cantidad de elementos por página
     * @param sortField Campo por el cual ordenar (por defecto 'id')
     * @param sortDirection Dirección del ordenamiento ('asc' o 'desc')
     * @returns Observable con la página de licencias solicitada
     */
    byPage(page: number, size: number, sortField: string = 'id', sortDirection: string = 'desc'): Observable<DataPackage> {
        return this.http.get<DataPackage>(
            encodeURI(`${this.licenciasUrl}/page?page=${page - 1}&size=${size}&sortField=${sortField}&sortDirection=${sortDirection}`)
        );
    }

    /**
     * Busca licencias que coincidan con el término de búsqueda.
     * 
     * @param searchTerm Término de búsqueda para filtrar licencias
     * @returns Observable con las licencias que coinciden con la búsqueda
     */
    search(searchTerm: string): Observable<DataPackage> {
        return this.http.get<DataPackage>(encodeURI(`${this.licenciasUrl}/search/${searchTerm}`));
    }

    /**
     * Obtiene el parte diario de licencias para una fecha específica
     * @param fecha Fecha para la cual se requiere el parte diario (formato: YYYY-MM-DD)
     * @returns Observable con el parte diario de licencias
     */
    getParteDiario(fecha: string): Observable<DataPackage> {
        return this.http.get<DataPackage>(encodeURI(`${this.licenciasUrl}/parte-diario/${fecha}`));
    }
}
